client/alloc_runner_test.go

```diff
@@ -232,24 +232,35 @@ func TestAllocRunner_DeploymentHealth_Unhealthy_Deadline(t *testing.T) {
 		if last == nil {
 			return false, fmt.Errorf("No updates")
 		}
+
+		// Assert alloc is unhealthy
 		if !last.DeploymentStatus.HasHealth() {
 			return false, fmt.Errorf("want deployment status unhealthy; got unset")
 		} else if *last.DeploymentStatus.Healthy {
 			return false, fmt.Errorf("want deployment status unhealthy; got healthy")
 		}
+
+		// Assert there is a task event explaining why we are unhealthy.
+		state, ok := last.TaskStates[task.Name]
+		if !ok {
+			return false, fmt.Errorf("missing state for task %s", task.Name)
+		}
+		n := len(state.Events)
+		if n == 0 {
+			return false, fmt.Errorf("no task events")
+		}
+		lastEvent := state.Events[n-1]
```

Comment:
Is the last event guaranteed to be the one this test is looking for? 

Wondering if it should scan all the events till it finds the "not running by deadline", rather than asserting this about the last event. Since events are sent and queued up async from the task runner, they may not always be in the same order..

Reply:
Good question and one I went back and forth on. I decided to leave it this way because we're looking for the failure event which should always be the last one. TaskEvents should be deterministically ordered even if _when_ the allocation updates/syncs occur is asynchronous.

I _think_ what was breaking it before is that restarts were enabled, so the test could see a _new_ `building task directory` event that was emitted by the restarted task *after* the first run fails.

Now that I disabled restarts in this test, the TaskEvents should be deterministic and eventually have `not running by deadline` as the final event.